visualization/app/codeCharta/util/nodeDecorator.ts

```diff
@@ -16,220 +15,210 @@ import { CodeMapHelper } from "./codeMapHelper"
+			// should be linear instead of O(n * m) [root.each() * node.leaves()] by
+			// using a depth first algorithm. Each current node tracks all
+			// `attributes`, `deltas` and `edgeAttributes` of the outer nodes
+			// combined. That way there's no need to traverse any node more than once.
+			//
+			// It is also possible to aggregate the stats linear. It's straight
+			// forward for non-median stats (the common case): these are just counted
+			// together. A map should be used to track the different depth.
+			// Calculating the median is more complex, both memory and computational
+			// wise. Instead of calculating the stats together of all outer nodes, an
+			// array of these should be collected (only if required!). New entries are
+			// then added to the array where they fit using array.splice. That way
+			// finding the median is linear instead of O(n log n). It is likely
+			// possible to optimize the latter further but that is probably too much
+			// work.
+			for (const { data } of node.leaves()) {
+				if (data.isExcluded) {
+					continue;
+				}
+				leafAttributes.push(data.attributes)
+				if (data.deltas !== undefined) {
+					leafDeltas.push(data.deltas)
+				}
+				if (!data.edgeAttributes) {
+					continue;
+				}
+				// Optimize for the common case
+				for (const [name, value] of Object.entries(data.edgeAttributes)) {
+					if (!value) {
+						continue;
+					}
+					if (attributeTypes.edges[name] === AttributeTypeValue.relative) {
+						// Use the median after collecting all entries.
+						if (edgeMetrics === undefined) {
+							edgeMetrics = {}
+						}
+						if (edgeMetrics[name] === undefined) {
+							edgeMetrics[name] = { incoming: [value.incoming], outgoing: [value.outgoing] }
+						} else {
+							edgeMetrics[name].incoming.push(value.incoming)
+							edgeMetrics[name].outgoing.push(value.outgoing)
+						}
+					} else {
+						// Set directly
+						node.data.edgeAttributes[name].incoming += value.incoming
+						node.data.edgeAttributes[name].outgoing += value.outgoing
+					}
+				}
+			}
+
+			for (const { name } of metricData) {
+				data.attributes[name] = this.aggregateLeafMetric(
+					leafAttributes,
+					name,
 					attributeTypes
 				)
 				if (isDeltaState) {
-					node.data.deltas[metric.name] = this.aggregateLeafMetric(
-						leaves.map(x => x.data.deltas),
-						metric.name,
+					data.deltas[name] = this.aggregateLeafMetric(
+						leafDeltas,
+						name,
 						attributeTypes
 					)
 				}
 			}
-		})
-	}
 
-	private static decorateNodeWithChildrenSumEdgeMetrics(
-		leaves: HierarchyNode<CodeMapNode>[],
-		node: HierarchyNode<CodeMapNode>,
-		edgeMetricData: NodeMetricData[],
-		attributeTypes: AttributeTypes
-	) {
-		edgeMetricData.forEach(edgeMetric => {
-			if (node.data.children && node.data.children.length > 0) {
-				node.data.edgeAttributes[edgeMetric.name] = this.aggregateLeafEdgeMetric(leaves, edgeMetric.name, attributeTypes)
+			if (edgeMetrics) {
```

Comment:
At this point edgeMetrics is always defined, isn't it?

Reply:
It's the lazy relative edge metric detection. If the node does not have those, they won't be defined.